File: Samples for Solid Edge/[C#]-Samples for Solid Edge/C#/SolidEdge.Draft.BatchPrint/CustomListView.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Windows.Forms;

namespace SolidEdge.Draft.BatchPrint
{
    public class CustomListView : ListView
    {
        private bool _useExplorerTheme;
        public DraftPrintUtilityOptions DraftPrintUtilityOptions;

        public CustomListView()
            : base()
        {
            // Turn on double buffering.
            DoubleBuffered = true;

            // Setup image list.
            SmallImageList = new ImageList();
            SmallImageList.ColorDepth = ColorDepth.Depth32Bit;
        }

        protected override void OnCreateControl()
        {
            base.OnCreateControl();

            // If set in designer, UseExplorerTheme will get called before control is actually created thus calling it a 2nd time here. 
            UseExplorerTheme = _useExplorerTheme;
        }

        protected override void OnDragOver(DragEventArgs e)
        {
            e.Effect = DragDropEffects.Copy;
        }

        protected override void OnDragDrop(DragEventArgs e)
        {
            // Can only drop folders\files, so check
            if (!e.Data.GetDataPresent(DataFormats.FileDrop)) 
            {
                return;
            }

            // Get the paths to the dropped items.
            string[] paths = (string[])e.Data.GetData(DataFormats.FileDrop);

            // Process each path.
            foreach (string path in paths)
            {
                // Determine if path is directory or file.
                if (Directory.Exists(path))
                {
                    SearchOption searchOption = SearchOption.TopDirectoryOnly;
                    DirectoryInfo directoryInfo = new DirectoryInfo(path);
                    DirectoryInfo[] subDirectoryInfos = directoryInfo.GetDirectories();

                    // If directory has subdirectories, ask user if we should process those as well.
                    if (subDirectoryInfos.Length > 0)
                    {
                        // Build the question to ask the user.
                        string message = String.Format("'{0}' contains subdirectories. Would you like to include those in the search?", directoryInfo.FullName);

                        // Ask the question.
                        switch (MessageBox.Show(message, "Process subdirectories?", MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question))
                        {
                            case DialogResult.Yes:
                                searchOption = SearchOption.AllDirectories;
                                break;
                            case DialogResult.Cancel:
                                // Bail out of entire OnDragDrop().
                                return;
                        }
                    }

                    AddFolder(directoryInfo, searchOption);
                }
                else
                {
                    FileInfo fileInfo = new FileInfo(path);
                    
                    // We are only interested in Draft files.
                    if (fileInfo.Extension.Equals(".dft", StringComparison.OrdinalIgnoreCase))
                    {
                        AddFile(fileInfo);
                    }
                }
            }

            // Adjust column(s) width to fit the item content.
            AutoResizeColumns(ColumnHeaderAutoResizeStyle.ColumnContent);
        }

        protected override void OnKeyDown(KeyEventArgs e)
        {
            base.OnKeyDown(e);

            try
            {
                // Disable ListView updating while adding items.
                BeginUpdate();

                // Determine the key pressed.
                switch (e.KeyCode)
                {
                    case Keys.A: // CTRL + A = Select All
                        if (e.Control)
                        {
                            foreach (ListViewItem item in Items)
                            {
                                item.Selected = true;
                            }
                        }
                        break;
                    case Keys.Delete: // DELETE = Remove selected items.
                        while (SelectedItems.Count > 0)
                        {
                            SelectedItems[0].Remove();
                        }
                        break;
                }
            }
            catch
            {
                throw;
            }
            finally
            {
                // Enable ListView updating.
                EndUpdate();
            }
        }

        public void AddFolder(DirectoryInfo directoryInfo)
        {
            AddFolder(directoryInfo, SearchOption.TopDirectoryOnly);
        }

        public void AddFolder(DirectoryInfo directoryInfo, SearchOption searchOption)
        {
            AddFiles(directoryInfo.GetFiles("*.dft", searchOption));
        }

        public void AddFiles(FileInfo[] fileInfos)
        {
            foreach (FileInfo fileInfo in fileInfos)
            {
                AddFile(fileInfo);
            }
        }

        public void AddFile(FileInfo fileInfo)
        {
            // Only add .dft files.
            if (!fileInfo.Extension.Equals(".dft", StringComparison.OrdinalIgnoreCase))
            {
                return;
            }

            // If the file has already been added, ignore it.
            if (Items.ContainsKey(fileInfo.FullName))
            {
                return;
            }

            // Add the icon to the image list if it's not there.
            if (!SmallImageList.Images.ContainsKey(fileInfo.Extension))
            {
                // Note: GetSmallIcon() is an extension method.
                Icon icon = fileInfo.GetSmallIcon();
                SmallImageList.Images.Add(fileInfo.Extension, icon);
            }

            
            // Add the list view item.
            ListViewItem listViewItem = Items.Add(
                key: fileInfo.FullName,
                text: fileInfo.FullName,
                imageKey: fileInfo.Extension);

            // Make a clone of the DraftPrintUtility settings for this individual item.
            listViewItem.Tag = DraftPrintUtilityOptions.Clone();
        }

        public bool UseExplorerTheme
        {
            get
            {
                return _useExplorerTheme;
            }
            set
            {
                _useExplorerTheme = value;
                if (this.Created)
                {
                    NativeMethods.SetWindowTheme(this.Handle, value ? "explorer" : "", null);
                }
            }
        }
    }
}

File: Samples for Solid Edge/[VB]-Samples for Solid Edge/VB/SolidEdge.Draft.BatchPrint/CustomListView.vb
Imports System.IO
Imports System.Runtime.InteropServices
Imports System.Text

Namespace SolidEdge.Draft.BatchPrint
	Public Class CustomListView
		Inherits ListView

		Private _useExplorerTheme As Boolean
		Public DraftPrintUtilityOptions As DraftPrintUtilityOptions

		Public Sub New()
			MyBase.New()
			' Turn on double buffering.
			DoubleBuffered = True

			' Setup image list.
			SmallImageList = New ImageList()
			SmallImageList.ColorDepth = ColorDepth.Depth32Bit
		End Sub

		Protected Overrides Sub OnCreateControl()
			MyBase.OnCreateControl()

			' If set in designer, UseExplorerTheme will get called before control is actually created thus calling it a 2nd time here. 
			UseExplorerTheme = _useExplorerTheme
		End Sub

		Protected Overrides Sub OnDragOver(ByVal e As DragEventArgs)
			e.Effect = DragDropEffects.Copy
		End Sub

		Protected Overrides Sub OnDragDrop(ByVal e As DragEventArgs)
			' Can only drop folders\files, so check
			If Not e.Data.GetDataPresent(DataFormats.FileDrop) Then
				Return
			End If

			' Get the paths to the dropped items.
			Dim paths() As String = CType(e.Data.GetData(DataFormats.FileDrop), String())

			' Process each path.
			For Each path As String In paths
				' Determine if path is directory or file.
				If Directory.Exists(path) Then
					Dim searchOption As SearchOption = System.IO.SearchOption.TopDirectoryOnly
					Dim directoryInfo As New DirectoryInfo(path)
					Dim subDirectoryInfos() As DirectoryInfo = directoryInfo.GetDirectories()

					' If directory has subdirectories, ask user if we should process those as well.
					If subDirectoryInfos.Length > 0 Then
						' Build the question to ask the user.
						Dim message As String = String.Format("'{0}' contains subdirectories. Would you like to include those in the search?", directoryInfo.FullName)

						' Ask the question.
						Select Case MessageBox.Show(message, "Process subdirectories?", MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question)
							Case DialogResult.Yes
								searchOption = System.IO.SearchOption.AllDirectories
							Case DialogResult.Cancel
								' Bail out of entire OnDragDrop().
								Return
						End Select
					End If

					AddFolder(directoryInfo, searchOption)
				Else
					Dim fileInfo As New FileInfo(path)

					' We are only interested in Draft files.
					If fileInfo.Extension.Equals(".dft", StringComparison.OrdinalIgnoreCase) Then
						AddFile(fileInfo)
					End If
				End If
			Next path

			' Adjust column(s) width to fit the item content.
			AutoResizeColumns(ColumnHeaderAutoResizeStyle.ColumnContent)
		End Sub

		Protected Overrides Sub OnKeyDown(ByVal e As KeyEventArgs)
			MyBase.OnKeyDown(e)

			Try
				' Disable ListView updating while adding items.
				BeginUpdate()

				' Determine the key pressed.
				Select Case e.KeyCode
					Case Keys.A ' CTRL + A = Select All
						If e.Control Then
							For Each item As ListViewItem In Items
								item.Selected = True
							Next item
						End If
					Case Keys.Delete ' DELETE = Remove selected items.
						Do While SelectedItems.Count > 0
							SelectedItems(0).Remove()
						Loop
				End Select
			Catch
				Throw
			Finally
				' Enable ListView updating.
				EndUpdate()
			End Try
		End Sub

		Public Sub AddFolder(ByVal directoryInfo As DirectoryInfo)
			AddFolder(directoryInfo, SearchOption.TopDirectoryOnly)
		End Sub

		Public Sub AddFolder(ByVal directoryInfo As DirectoryInfo, ByVal searchOption As SearchOption)
			AddFiles(directoryInfo.GetFiles("*.dft", searchOption))
		End Sub

		Public Sub AddFiles(ByVal fileInfos() As FileInfo)
			For Each fileInfo As FileInfo In fileInfos
				AddFile(fileInfo)
			Next fileInfo
		End Sub

		Public Sub AddFile(ByVal fileInfo As FileInfo)
			' Only add .dft files.
			If Not fileInfo.Extension.Equals(".dft", StringComparison.OrdinalIgnoreCase) Then
				Return
			End If

			' If the file has already been added, ignore it.
			If Items.ContainsKey(fileInfo.FullName) Then
				Return
			End If

			' Add the icon to the image list if it's not there.
			If Not SmallImageList.Images.ContainsKey(fileInfo.Extension) Then
				' Note: GetSmallIcon() is an extension method.
				Dim icon As Icon = fileInfo.GetSmallIcon()
				SmallImageList.Images.Add(fileInfo.Extension, icon)
			End If


			' Add the list view item.
			Dim listViewItem As ListViewItem = Items.Add(key:= fileInfo.FullName, text:= fileInfo.FullName, imageKey:= fileInfo.Extension)

			' Make a clone of the DraftPrintUtility settings for this individual item.
			listViewItem.Tag = DraftPrintUtilityOptions.Clone()
		End Sub

		Public Property UseExplorerTheme() As Boolean
			Get
				Return _useExplorerTheme
			End Get
			Set(ByVal value As Boolean)
				_useExplorerTheme = value
				If Me.Created Then
					NativeMethods.SetWindowTheme(Me.Handle,If(value, "explorer", ""), Nothing)
				End If
			End Set
		End Property
	End Class
End Namespace
